lib/Doctrine/Migrations/VersionExecutor.php

```diff
@@ -0,0 +1,323 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Migrations;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\Migrations\Configuration\Configuration;
+use Doctrine\Migrations\Exception\SkipMigration;
+use Doctrine\Migrations\Provider\SchemaDiffProviderInterface;
+use Throwable;
+use function count;
+use function is_array;
+use function microtime;
+use function round;
+use function rtrim;
+use function sprintf;
+use function ucfirst;
+
+final class VersionExecutor
+{
+    /** @var Configuration */
+    private $configuration;
+
+    /** @var Connection */
+    private $connection;
+
+    /** @var SchemaDiffProviderInterface */
+    private $schemaProvider;
+
+    /** @var OutputWriter */
+    private $outputWriter;
+
+    /** @var ParameterFormatter */
+    private $parameterFormatter;
+
+    /** @var string[] */
+    private $sql = [];
+
+    /** @var mixed[] */
+    private $params = [];
+
+    /** @var mixed[] */
+    private $types = [];
+
+    public function __construct(
+        Configuration $configuration,
+        Connection $connection,
+        SchemaDiffProviderInterface $schemaProvider,
+        OutputWriter $outputWriter
+    ) {
+        $this->configuration  = $configuration;
+        $this->connection     = $connection;
+        $this->schemaProvider = $schemaProvider;
+        $this->outputWriter   = $outputWriter;
+
+        $this->parameterFormatter = new ParameterFormatter($this->connection);
+    }
+
+    /**
+     * @param string[]|string $sql
```

Comment:
I'd try to eliminate this by always passing an array of strings (or a collection). This leads to some nasty if-else tricks.

Reply:
I can make this change here, I won't be able to make the change in the addSql() that the end user uses as it would be a big BC break, but I can make it so the internal addSql here always receives a `string[]`